subsys/bluetooth/mesh/dfu_slot.c

```diff
@@ -147,146 +122,172 @@ bt_mesh_dfu_slot_add(size_t size, const uint8_t *fwid,
 		return NULL;
 	}
 
-	slot->slot.fwid_len = fwid_len;
-	slot->slot.metadata_len = metadata_len;
-	slot->slot.uri_len = uri_len;
-	memcpy(slot->slot.fwid, fwid, fwid_len);
-	memcpy(slot->slot.metadata, metadata, metadata_len);
-	memcpy(slot->slot.uri, uri, uri_len);
-	slot->slot.size = size;
+	slot->slot.fwid_len = 0;
+	slot->slot.metadata_len = 0;
+	slot->slot.size = 0;
+	slot->state = SLOT_STATE_RESERVED;
 
-	err = slot_store(slot);
-	if (err) {
-		slot_invalidate(slot);
-		LOG_WRN("Store failed (err: %d)", err);
-		return NULL;
-	}
-
-	sys_slist_append(&list, &slot->n);
+	LOG_DBG("Reserved slot #%u", slot - &slots[0]);
 
-	LOG_DBG("Added slot #%u: %s", slot - &slots[0],
-		bt_hex(slot->slot.fwid, slot->slot.fwid_len));
 	return &slot->slot;
 }
 
-int bt_mesh_dfu_slot_valid_set(const struct bt_mesh_dfu_slot *slot, bool valid)
+int bt_mesh_dfu_slot_info_set(struct bt_mesh_dfu_slot *dfu_slot, size_t size,
+			      const uint8_t *metadata, size_t metadata_len)
 {
-	uint16_t idx;
-	bool prev;
-	int err;
+	struct slot *slot = CONTAINER_OF(dfu_slot, struct slot, slot);
 
-	if (!SLOT_IN_ARRAY(slot) || !slot_in_use(slot)) {
-		return -ENOENT;
+	if (metadata_len > CONFIG_BT_MESH_DFU_METADATA_MAXLEN) {
+		return -EFBIG;
 	}
 
-	idx = slot_idx(slot);
+	if (slot->state != SLOT_STATE_RESERVED) {
+		return -EINVAL;
+	}
 
-	LOG_DBG("%u: %u", idx, valid);
+	slot->slot.size = size;
+	slot->slot.metadata_len = metadata_len;
+	memcpy(slot->slot.metadata, metadata, metadata_len);
+	return 0;
+}
 
-	if (valid) {
-		prev = atomic_test_and_set_bit(valid_slots, idx);
-	} else {
-		prev = atomic_test_and_clear_bit(valid_slots, idx);
-	}
+int bt_mesh_dfu_slot_fwid_set(struct bt_mesh_dfu_slot *dfu_slot,
+			      const uint8_t *fwid, size_t fwid_len)
+{
+	struct slot *slot = CONTAINER_OF(dfu_slot, struct slot, slot);
 
-	if (valid == prev) {
-		return 0;
+	if (fwid_len > CONFIG_BT_MESH_DFU_FWID_MAXLEN) {
+		return -EFBIG;
 	}
 
-	err = valid_slots_store();
-	if (err) {
-		LOG_WRN("Storage failed. err: %d", err);
-		atomic_set_bit_to(valid_slots, idx, prev);
+	if (slot->state != SLOT_STATE_RESERVED) {
+		return -EINVAL;
 	}
 
-	return err;
+	slot->slot.fwid_len = fwid_len;
+	memcpy(slot->slot.fwid, fwid, fwid_len);
+	return 0;
 }
 
-bool bt_mesh_dfu_slot_is_valid(const struct bt_mesh_dfu_slot *slot)
+int bt_mesh_dfu_slot_commit(struct bt_mesh_dfu_slot *dfu_slot)
 {
-	uint16_t idx;
+	int err;
+	struct slot *slot = CONTAINER_OF(dfu_slot, struct slot, slot);
+
+	if (slot->state != SLOT_STATE_RESERVED ||
+	    slot->slot.fwid_len == 0 ||
+	    slot->slot.size == 0) {
+		return -EINVAL;
+	}
+
+	if (bt_mesh_dfu_slot_get(slot->slot.fwid, slot->slot.fwid_len, NULL)
+	    >= 0) {
+		return -EALREADY;
+	}
```

Comment:
It seems it is already too late to check this. `_commit` is called after a fw is received which is usually a long process. We should not allow starting uploading two identical firmware ids. We will also ideally need to check this error code in dfd_srv.c, but currently we just ignore it: https://github.com/zephyrproject-rtos/zephyr/blob/702ec88e3c9da13932b6780ab0ef3624408bc344/subsys/bluetooth/mesh/dfd_srv.c#L828
From what I see uniqueness of the fwid is already checked in dfd_srv.c in `bt_mesh_dfd_srv_oob_check_complete` and in `handle_upload_start`, so it looks redundant. Since dfu slot api is internal, it should be safe to remove this check.

Reply:
This is here because it's legal to have two DFD servers on the same node, these will share the image list, and the `_check_complete` will pass for both as long as none of them have commited yet (it only checks for a _commited_ slot with fwid == fwid). This check is to avoid the race condition where both pass the check.

Reply:
Neither of both DFD servers should be able to start uploading same firmware. So this check should be done against reserved slots then.

Reply:
Right, then I'll have to change my design a bit. Currently, reserved slots are "invisible" to all other APIs so you can't tell if a non-commited slot exists that has the same fwid. But I see your point, better to get feedback early to not have to unnecessarily download the same image twice.